File: shoring/compat/src/main/java/com/panopset/compat/Dateop.java
package com.panopset.compat;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Date;

public class Dateop {
	public static ZoneId ZONE_ID = ZoneId.of("America/Los_Angeles");

	public static String show(LocalDate date) {
		return YMD_FMT.format(date);
	}

	public static LocalDate parse(String source) {
		return LocalDate.parse(source.trim(), YMD_FMT);
	}

	private static final DateTimeFormatter YMD_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd");

	public static String fill2(int val) {
		return val < 10 ? "0" + val : "" + val;
	}

	public static String createDateKey(LocalDate ld) {
		return createDateKey(ld.getYear(), ld.getMonthValue(), ld.getDayOfMonth());
	}

	public static String createMonthKey(LocalDate ld) {
		return createMonthKey(ld.getYear(), ld.getMonthValue());
	}

	public static String createDateKey(int yr, int mn, int dy) {
		return String.format("%s-%s", createMonthKey(yr, mn), Dateop.fill2(dy));
	}

	public static String createMonthKey(int yr, int mn) {
		return String.format("%d-%s", yr, Dateop.fill2(mn));
	}

	public static Date toPacific(LocalDate oldest) {
		// http://stackoverflow.com/questions/33066904
		return Date.from(oldest.atStartOfDay(ZONE_ID).toInstant());
	}

	public static LocalDate toLocalDate(Date date) {
		// http://stackoverflow.com/questions/21242110
		return Instant.ofEpochMilli(date.getTime()).atZone(ZONE_ID).toLocalDate();
	}
}

File: code/shoring/compat/src/main/kotlin/com/panopset/compat/Dateop.kt
package com.panopset.compat

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.*

object Dateop {
    var ZONE_ID = ZoneId.of("America/Los_Angeles")
    fun show(date: LocalDate?): String {
        return YMD_FMT.format(date)
    }

    fun parse(source: String): LocalDate {
        return LocalDate.parse(source.trim { it <= ' ' }, YMD_FMT)
    }

    private val YMD_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd")
    private fun fill2(value: Int): String {
        return if (value < 10) "0$value" else "" + value
    }

    fun createDateKey(ld: LocalDate): String {
        return createDateKey(ld.year, ld.monthValue, ld.dayOfMonth)
    }

    fun createMonthKey(ld: LocalDate): String {
        return createMonthKey(ld.year, ld.monthValue)
    }

    private fun createDateKey(yr: Int, mn: Int, dy: Int): String {
        return String.format("%s-%s", createMonthKey(yr, mn), fill2(dy))
    }

    private fun createMonthKey(yr: Int, mn: Int): String {
        return String.format("%d-%s", yr, fill2(mn))
    }

    fun toPacific(oldest: LocalDate): Date {
        // http://stackoverflow.com/questions/33066904
        return Date.from(oldest.atStartOfDay(ZONE_ID).toInstant())
    }

    fun toLocalDate(date: Date): LocalDate {
        // http://stackoverflow.com/questions/21242110
        return Instant.ofEpochMilli(date.time).atZone(ZONE_ID).toLocalDate()
    }
}
